<a href="https://colab.research.google.com/github/Alifhidayatunnisa/Basic-Python-Programming-for-Data-Science/blob/main/DataFrame.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Frame Pada Python

DataFrame berbentuk dua dimensi sebagaimana matriks. DataFrame adalah struktur data yang memungkinkan untuk menyimpan beberapa variabel dengan lebih dari satu tipe data. Data dalam satu variabel yang sama disimpan dalam satu kolom yang dapat diberi nama. DataFrame adalah representasi khusus untuk mengolah data dalam bentuk spreadsheet. Hal ini memungkinkan kita untuk mengimpor file dari ekstensi CSV, XLS, atau lainnya ke dalam Python.

### Mengimpor Library

In [ ]:
import pandas as pd
import numpy as np

In [1]:
import pandas as pd
import numpy as np

### Membuat Data Frame Pada Python

**Membuat Data Frame dari List**

In [ ]:
# Membuat List isian dataframe
List1 = ['Jambu', 'Apel', 'Semangka', 'Anggur', 'Ceplukan', 'Saradan', 'Salak']
  
# Membuat DataFrame
df = pd.DataFrame(List1)

# Memanggil DataFrame
df

0
0     Jambu
1      Apel
2  Semangka
3    Anggur
4  Ceplukan
5   Saradan
6     Salak

In [2]:
List1 = ['Thai Tea Bobba', 'Matcha Bobba Milk', 'Brown Sugar Bobba Milk', 'Choco Bobba Milk', 'Strawberry Bobba Milk']
df = pd.DataFrame(List1)
df

0
0          Thai Tea Bobba
1       Matcha Bobba Milk
2  Brown Sugar Bobba Milk
3        Choco Bobba Milk
4   Strawberry Bobba Milk

In [ ]:
# Membuat List isian DataFrame
List1 =  ['Jambu', 'Apel', 'Semangka', 'Anggur', 'Ceplukan', 'Saradan', 'Salak']

# Membuat DataFrame dengan Indeks dan Nama Kolom Tertentu
df = pd.DataFrame(List1,
                  index=["Satu", "Dua", "Tiga", "Empat", "Lima", "Enam", "Tujuh"], # Menamai baris
                  columns=["Nama Buah"])                                         # Menamai kolom

# Memanggil DataFrame
df

Nama Buah
Satu      Jambu
Dua        Apel
Tiga   Semangka
Empat    Anggur
Lima   Ceplukan
Enam    Saradan
Tujuh     Salak

In [3]:
List1 = ['Thai Tea Bobba', 'Matcha Bobba Milk', 'Brown Sugar Bobba Milk', 'Choco Bobba Milk', 'Strawberry Bobba Milk']
df = pd.DataFrame(List1,
                  index=["Satu", "Dua", "Tiga", "Empat", "Lima"],
                  columns=["Menu"])
df

Menu
Satu           Thai Tea Bobba
Dua         Matcha Bobba Milk
Tiga   Brown Sugar Bobba Milk
Empat        Choco Bobba Milk
Lima    Strawberry Bobba Milk

In [ ]:
# Membuat List isian DataFrame
List1 = ['Jambu', 'Apel', 'Semangka', 'Anggur', 'Ceplukan', 'Saradan', 'Salak']
List2 = [3000, 5000, 5000, 7000, 2500, 8000, 4000]

# Membuat DataFrame dengan Indeks dan Nama Kolom Tertentu
df = pd.DataFrame(list(zip(List1, List2)),
                  columns=["Nama Buah", "Harga Per Kg"])  # Menamai Kolom
                  
#Memanggil DataFrame
df

Nama Buah  Harga Per Kg
0     Jambu          3000
1      Apel          5000
2  Semangka          5000
3    Anggur          7000
4  Ceplukan          2500
5   Saradan          8000
6     Salak          4000

In [4]:
List1 = ['Thai Tea Bobba', 'Matcha Bobba Milk', 'Brown Sugar Bobba Milk', 'Choco Bobba Milk', 'Strawberry Bobba Milk']
List2 = [10000, 12000, 11000, 13000, 12000]
df = pd.DataFrame(list(zip(List1, List2)),
                  columns=["Menu", "Harga"])
df

Menu  Harga
0          Thai Tea Bobba  10000
1       Matcha Bobba Milk  12000
2  Brown Sugar Bobba Milk  11000
3        Choco Bobba Milk  13000
4   Strawberry Bobba Milk  12000

In [ ]:
# Membuat List MultiDimensi     
List1 = [['Haikal', 21], ['Indra', 20],
       ['Mujhid', 22], ['Satria', 21]]

# Membuat Data Frame
df = pd.DataFrame(List1, columns =['Nama', 'Umur'])

#Memanggil Data Frame
df

Nama  Umur
0  Haikal    21
1   Indra    20
2  Mujhid    22
3  Satria    21

In [5]:
List3 = [['Nisa', 12], ['Keke', 7], ['Nando', 3], ['Bella', 6], ['Zehra', 1]]
df = pd.DataFrame(List3, columns = ["Nama", "Kelas"])
df

Nama  Kelas
0   Nisa     12
1   Keke      7
2  Nando      3
3  Bella      6
4  Zehra      1

**Membuat Data Frame dari Dictionary**


In [ ]:
# Membuat List untuk Dictionary
List_Nama    = ["Tumiyem", "Semi", "Pairah", "Juminem"]
List_Jurusan = ["Matematika", "Akuntansi", "Mikrobiologi", "Aktuaria"]
List_Nilai   = [90, 40, 80, 98]
  
# Membuat Dictionary
dict1 = {'Nama': List_Nama, 'Juruan': List_Jurusan, 'Nilai': List_Nilai} 

# Membuat Data Frame
df = pd.DataFrame(dict1)

# Memanggil Data Frame
df

Nama        Juruan  Nilai
0  Tumiyem    Matematika     90
1     Semi     Akuntansi     40
2   Pairah  Mikrobiologi     80
3  Juminem      Aktuaria     98

In [6]:
List_Nama = ["IU", "Jennie", "Irene", "Nayeon", "Yoona"]
List_Agensi = ['EDAM', 'YG', 'SM', 'JYP', 'SM']
List_Umur = [27, 25, 30, 25, 30]
dict1 = {'Nama' : List_Nama, 'Agensi' : List_Agensi, 'Umur' : List_Umur}
df = pd.DataFrame(dict1)
df

Nama Agensi  Umur
0      IU   EDAM    27
1  Jennie     YG    25
2   Irene     SM    30
3  Nayeon    JYP    25
4   Yoona     SM    30

### Mengimpor DataFrame Pada Python

**Mengimpor Data Frame dari Format CSV**

In [ ]:
# Mengimpor Data Frame
df = pd.read_csv("https://raw.githubusercontent.com/jokoeliyanto/Kelas-Dasar-Pejuang-Data-2.0/main/Super-Store-Dataset.csv")

# Memanggil Data Frame
df

order_id  order_date customer_id  ...     sales quantity    profit
0     CA-2016-152156   11/8/2016    CG-12520  ...  261.9600        2   41.9136
1     CA-2016-152156   11/8/2016    CG-12520  ...  731.9400        3  219.5820
2     CA-2016-138688   6/12/2016    DV-13045  ...   14.6200        2    6.8714
3     US-2015-108966  10/11/2015    SO-20335  ...  957.5775        5 -383.0310
4     US-2015-108966  10/11/2015    SO-20335  ...   22.3680        2    2.5164
...              ...         ...         ...  ...       ...      ...       ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...   25.2480        3    4.1028
9984  CA-2017-121258   2/26/2017    DB-13060  ...   91.9600        2   15.6332
9985  CA-2017-121258   2/26/2017    DB-13060  ...  258.5760        2   19.3932
9986  CA-2017-121258   2/26/2017    DB-13060  ...   29.6000        4   13.3200
9987  CA-2017-119914    5/4/2017    CC-12220  ...  243.1600        2   72.9480

[9988 rows x 9 columns]

In [11]:
df = pd.read_csv("https://raw.githubusercontent.com/Alifhidayatunnisa/Basic-Python-Programming-for-Data-Science/main/Super%20Store%20Dataset.csv")
df

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9988 rows x 9 columns]

**Mengimpor Data Frame dari Format XLS/XLSX**

In [ ]:
# Mengimpor Data Frame
df = pd.read_excel("/content/SuperStoreDataset.xlsx") #Pastikan sudah diupload di folder

# Memanggil Data Frame
df

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9988 rows x 9 columns]

In [17]:
df = pd.read_excel("/content/Super Store Dataset.xlsx")
df

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9988 rows x 9 columns]

### Eksplorasi Data Frame

**Head**

In [ ]:
#Melihat data teratas pada dataframe(default 5)
df.head()

order_id  order_date customer_id  ...    sales quantity  profit
0  CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1  CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2  CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3  US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4  US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164

[5 rows x 9 columns]

In [18]:
df.head()

order_id  order_date customer_id  ...    sales quantity  profit
0  CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1  CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2  CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3  US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4  US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164

[5 rows x 9 columns]

**Tails**

In [ ]:
#Melihat data terbawah pada dataframe(default 5)
df.tail()

order_id order_date customer_id  ...   sales quantity  profit
9983  CA-2014-110422  1/21/2014    TB-21400  ...   25248        3   41028
9984  CA-2017-121258  2/26/2017    DB-13060  ...    9196        2  156332
9985  CA-2017-121258  2/26/2017    DB-13060  ...  258576        2  193932
9986  CA-2017-121258  2/26/2017    DB-13060  ...     296        4    1332
9987  CA-2017-119914   5/4/2017    CC-12220  ...   24316        2   72948

[5 rows x 9 columns]

In [19]:
df.tail()

order_id order_date customer_id  ...   sales quantity  profit
9983  CA-2014-110422  1/21/2014    TB-21400  ...   25248        3   41028
9984  CA-2017-121258  2/26/2017    DB-13060  ...    9196        2  156332
9985  CA-2017-121258  2/26/2017    DB-13060  ...  258576        2  193932
9986  CA-2017-121258  2/26/2017    DB-13060  ...     296        4    1332
9987  CA-2017-119914   5/4/2017    CC-12220  ...   24316        2   72948

[5 rows x 9 columns]

**Shape**

In [ ]:
#Melihat ukuran data frame(Baris X Kolom)
df.shape

(9988, 9)

In [20]:
df.shape

(9988, 9)

**Column Names**

In [ ]:
#Melihat nama-nama Kolom
df.columns.values 

array(['order_id', 'order_date', 'customer_id', 'segment', 'category',
       'sub_category', 'sales', 'quantity', 'profit'], dtype=object)

In [21]:
df.columns.values 

array(['order_id', 'order_date', 'customer_id', 'segment', 'category',
       'sub_category', 'sales', 'quantity', 'profit'], dtype=object)

**Tipe Data Masing-masing Kolom**

In [ ]:
#Melihat Tipe Data Masing-masing Kolom
df.dtypes

order_id        object
order_date      object
customer_id     object
segment         object
category        object
sub_category    object
sales            int64
quantity         int64
profit           int64
dtype: object

In [22]:
df.dtypes

order_id        object
order_date      object
customer_id     object
segment         object
category        object
sub_category    object
sales            int64
quantity         int64
profit           int64
dtype: object

In [ ]:
#Melihat Tipe Data Masing-masing Kolom dengan lebih lengkap
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9988 entries, 0 to 9987
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   order_id      9988 non-null   object
 1   order_date    9988 non-null   object
 2   customer_id   9988 non-null   object
 3   segment       9988 non-null   object
 4   category      9988 non-null   object
 5   sub_category  9988 non-null   object
 6   sales         9988 non-null   int64 
 7   quantity      9988 non-null   int64 
 8   profit        9988 non-null   int64 
dtypes: int64(3), object(6)
memory usage: 702.4+ KB


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9988 entries, 0 to 9987
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   order_id      9988 non-null   object
 1   order_date    9988 non-null   object
 2   customer_id   9988 non-null   object
 3   segment       9988 non-null   object
 4   category      9988 non-null   object
 5   sub_category  9988 non-null   object
 6   sales         9988 non-null   int64 
 7   quantity      9988 non-null   int64 
 8   profit        9988 non-null   int64 
dtypes: int64(3), object(6)
memory usage: 702.4+ KB


**Describe**

In [ ]:
#Melihat statistik masing-masing kolom bertipe numerik
df.describe()

sales     quantity        profit
count  9.988000e+03  9988.000000  9.988000e+03
mean   1.110761e+05     3.789147  1.661381e+05
std    5.038903e+05     2.225586  1.438964e+06
min    3.000000e+00     1.000000 -3.839990e+07
25%    2.512000e+03     2.000000  2.245000e+03
50%    1.102150e+04     3.000000  3.629500e+04
75%    5.390700e+04     5.000000  1.509840e+05
max    2.396266e+07    14.000000  4.630476e+07

In [24]:
df.describe()

sales     quantity        profit
count  9.988000e+03  9988.000000  9.988000e+03
mean   1.110761e+05     3.789147  1.661381e+05
std    5.038903e+05     2.225586  1.438964e+06
min    3.000000e+00     1.000000 -3.839990e+07
25%    2.512000e+03     2.000000  2.245000e+03
50%    1.102150e+04     3.000000  3.629500e+04
75%    5.390700e+04     5.000000  1.509840e+05
max    2.396266e+07    14.000000  4.630476e+07

### Akses Data Frame(Slicing)

**Mengakses DataFrame Berdasarkan Baris**

In [ ]:
df[0:10] #Mengakses data Frame dari baris pertama sampai baris ke 10(indeks 0 sampai indeks sebelum 10)

order_id  order_date customer_id  ...    sales quantity  profit
0  CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1  CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2  CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3  US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4  US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
5  CA-2014-115812    6/9/2014    BH-11710  ...     4886        7  141694
6  CA-2014-115812    6/9/2014    BH-11710  ...      728        4   19656
7  CA-2014-115812    6/9/2014    BH-11710  ...   907152        6  907152
8  CA-2014-115812    6/9/2014    BH-11710  ...    18504        3   57825
9  CA-2014-115812    6/9/2014    BH-11710  ...     1149        5    3447

[10 rows x 9 columns]

In [25]:
df[0:7]

order_id  order_date customer_id  ...    sales quantity  profit
0  CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1  CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2  CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3  US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4  US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
5  CA-2014-115812    6/9/2014    BH-11710  ...     4886        7  141694
6  CA-2014-115812    6/9/2014    BH-11710  ...      728        4   19656

[7 rows x 9 columns]

**Mengakses DataFrame Berdasarkan Nama Kolom**

In [ ]:
#Mengakses 1 kolom pada Data Frame, Hasilnya berupa Series
df['profit'] 

0       419136
1       219582
2        68714
3      -383031
4        25164
         ...  
9983     41028
9984    156332
9985    193932
9986      1332
9987     72948
Name: profit, Length: 9988, dtype: int64

In [26]:
df['category'] 

0             Furniture
1             Furniture
2       Office Supplies
3             Furniture
4       Office Supplies
             ...       
9983          Furniture
9984          Furniture
9985         Technology
9986    Office Supplies
9987    Office Supplies
Name: category, Length: 9988, dtype: object

In [ ]:
#[Alternatif lain] Mengakses 1 kolom pada Data Frame, Hasilnya berupa Series
df.profit 

0       419136
1       219582
2        68714
3      -383031
4        25164
         ...  
9983     41028
9984    156332
9985    193932
9986      1332
9987     72948
Name: profit, Length: 9988, dtype: int64

In [27]:
df.category

0             Furniture
1             Furniture
2       Office Supplies
3             Furniture
4       Office Supplies
             ...       
9983          Furniture
9984          Furniture
9985         Technology
9986    Office Supplies
9987    Office Supplies
Name: category, Length: 9988, dtype: object

In [ ]:
#Mengakses Data Frame lebih dari 1 kolom
df[['profit', 'customer_id']] 

profit customer_id
0     419136    CG-12520
1     219582    CG-12520
2      68714    DV-13045
3    -383031    SO-20335
4      25164    SO-20335
...      ...         ...
9983   41028    TB-21400
9984  156332    DB-13060
9985  193932    DB-13060
9986    1332    DB-13060
9987   72948    CC-12220

[9988 rows x 2 columns]

In [28]:
df[['category', 'quantity']]

category  quantity
0           Furniture         2
1           Furniture         3
2     Office Supplies         2
3           Furniture         5
4     Office Supplies         2
...               ...       ...
9983        Furniture         3
9984        Furniture         2
9985       Technology         2
9986  Office Supplies         4
9987  Office Supplies         2

[9988 rows x 2 columns]

In [ ]:
#Menerapkan fungsi head pada slicing data frame
df[['profit', 'customer_id']].head(6)

profit customer_id
0  419136    CG-12520
1  219582    CG-12520
2   68714    DV-13045
3 -383031    SO-20335
4   25164    SO-20335
5  141694    BH-11710

In [29]:
df[['category', 'quantity']].head(7)

category  quantity
0        Furniture         2
1        Furniture         3
2  Office Supplies         2
3        Furniture         5
4  Office Supplies         2
5        Furniture         7
6  Office Supplies         4

In [ ]:
#Mengakses data frame pada baris dan kolom tertentu
df.iloc[0,2]

'CG-12520'

In [32]:
df.iloc[2,2]

'DV-13045'

In [ ]:
#Mengakses data frame pada baris dan kolom tertentu [Bentuk Interval]
df.iloc[0:10,2:5]

customer_id    segment         category
0    CG-12520   Consumer        Furniture
1    CG-12520   Consumer        Furniture
2    DV-13045  Corporate  Office Supplies
3    SO-20335   Consumer        Furniture
4    SO-20335   Consumer  Office Supplies
5    BH-11710   Consumer        Furniture
6    BH-11710   Consumer  Office Supplies
7    BH-11710   Consumer       Technology
8    BH-11710   Consumer  Office Supplies
9    BH-11710   Consumer  Office Supplies

In [33]:
df.iloc[0:7,4:7]

category sub_category    sales
0        Furniture    Bookcases    26196
1        Furniture       Chairs    73194
2  Office Supplies       Labels     1462
3        Furniture       Tables  9575775
4  Office Supplies      Storage    22368
5        Furniture  Furnishings     4886
6  Office Supplies          Art      728

In [ ]:
#Mengakses data pada kolom tertentu dengan kriteria tertentu
print(df[df['profit']>=68714])

print(df['profit'].max())

            order_id  order_date customer_id  ...   sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...   26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...   73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...    1462        2   68714
5     CA-2014-115812    6/9/2014    BH-11710  ...    4886        7  141694
7     CA-2014-115812    6/9/2014    BH-11710  ...  907152        6  907152
...              ...         ...         ...  ...     ...      ...     ...
9980  CA-2016-125794   9/29/2016    ML-17410  ...    3624        1  152208
9981  CA-2017-163629  11/17/2017    RA-19885  ...    7999        1  287964
9984  CA-2017-121258   2/26/2017    DB-13060  ...    9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...  258576        2  193932
9987  CA-2017-119914    5/4/2017    CC-12220  ...   24316        2   72948

[3854 rows x 9 columns]
46304755


In [34]:
print(df[df['sales']>=70000])

print(df['sales'].max())

            order_id  order_date customer_id  ...    sales quantity   profit
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3   219582
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5  -383031
7     CA-2014-115812    6/9/2014    BH-11710  ...   907152        6   907152
10    CA-2014-115812    6/9/2014    BH-11710  ...  1706184        9   853092
11    CA-2014-115812    6/9/2014    BH-11710  ...   911424        4   683568
...              ...         ...         ...  ...      ...      ...      ...
9962  CA-2017-153871  12/11/2017    RB-19435  ...    73598        2   331191
9966  CA-2016-130225   9/11/2016    RC-19960  ...    99568        2   336042
9970  US-2016-103674   12/6/2016    AP-10720  ...   249584        2    31198
9973  US-2016-103674   12/6/2016    AP-10720  ...   437472       14  1531152
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2   193932

[2152 rows x 9 columns]
23962656


In [ ]:
#Memperoleh indeks hasil pencarian data dengan kriteria tertentu
max_profit=df['profit'].max()
df.index[df['profit'] == max_profit].tolist()

[4096]

In [35]:
max_sales=df['sales'].max()
df.index[df['sales'] == max_sales].tolist()

[399]

In [ ]:
#Memperoleh indeks hasil pencarian data dengan kriteria tertentu (Menggunakan And)
df[(df['profit']>=68714) & (df['quantity']>=2)]

order_id order_date customer_id  ...   sales quantity  profit
0     CA-2016-152156  11/8/2016    CG-12520  ...   26196        2  419136
1     CA-2016-152156  11/8/2016    CG-12520  ...   73194        3  219582
2     CA-2016-138688  6/12/2016    DV-13045  ...    1462        2   68714
5     CA-2014-115812   6/9/2014    BH-11710  ...    4886        7  141694
7     CA-2014-115812   6/9/2014    BH-11710  ...  907152        6  907152
...              ...        ...         ...  ...     ...      ...     ...
9976  US-2016-157728  9/22/2016    RC-19960  ...    3556        7  167132
9977  US-2016-157728  9/22/2016    RC-19960  ...    9798        2  274344
9984  CA-2017-121258  2/26/2017    DB-13060  ...    9196        2  156332
9985  CA-2017-121258  2/26/2017    DB-13060  ...  258576        2  193932
9987  CA-2017-119914   5/4/2017    CC-12220  ...   24316        2   72948

[3654 rows x 9 columns]

In [36]:
df[(df['sales']>=70000) & (df['profit']>=60000)]

order_id  order_date customer_id  ...    sales quantity   profit
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3   219582
7     CA-2014-115812    6/9/2014    BH-11710  ...   907152        6   907152
10    CA-2014-115812    6/9/2014    BH-11710  ...  1706184        9   853092
11    CA-2014-115812    6/9/2014    BH-11710  ...   911424        4   683568
13    CA-2016-161389   12/5/2016    IM-15070  ...   407976        3  1325922
...              ...         ...         ...  ...      ...      ...      ...
9949  CA-2015-141593  12/14/2015    DB-12970  ...   273568        2   102588
9962  CA-2017-153871  12/11/2017    RB-19435  ...    73598        2   331191
9966  CA-2016-130225   9/11/2016    RC-19960  ...    99568        2   336042
9973  US-2016-103674   12/6/2016    AP-10720  ...   437472       14  1531152
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2   193932

[1156 rows x 9 columns]

In [ ]:
#Memperoleh indeks hasil pencarian data dengan kriteria tertentu (Menggunakan Or)
df[(df['profit']>=68714) | (df['quantity']>=2)]

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9289 rows x 9 columns]

In [37]:
df[(df['sales']>=70000) | (df['profit']>=60000)]

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
5     CA-2014-115812    6/9/2014    BH-11710  ...     4886        7  141694
...              ...         ...         ...  ...      ...      ...     ...
9980  CA-2016-125794   9/29/2016    ML-17410  ...     3624        1  152208
9981  CA-2017-163629  11/17/2017    RA-19885  ...     7999        1  287964
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[5113 rows x 9 columns]

### Modifikasi Data Frame 1(Sort & Cleaning)

**Mengurutkan Data Berdasarkan Kolom tertentu**

In [ ]:
#Mengurutkan data frame berdasarkan 1 kolom
df.sort_values(['profit'], ascending=True)

order_id  order_date customer_id  ...    sales quantity    profit
683   US-2017-168116   11/4/2017    GT-14635  ...   799998        4 -38399904
9768  CA-2014-169019   7/26/2014    LF-17185  ...  2177584        8 -37018928
4988  US-2017-122714   12/7/2017    HG-14965  ...   188999        5 -29294845
3149  CA-2015-147830  12/15/2015    NF-18385  ...  1799994        2 -26399912
9633  CA-2015-116638   1/28/2015    JH-15985  ...  4297644       13 -18623124
...              ...         ...         ...  ...      ...      ...       ...
4274  US-2016-107440   4/16/2016    BS-11365  ...   909993        7  23659818
1085  US-2016-143819    3/1/2016    KD-16270  ...   489993        7  24009657
6517  CA-2017-138289   1/16/2017    AR-10540  ...   544396        4  25042216
7662  US-2016-140158   10/4/2016    DR-12940  ...   539991        9  25919568
4096  CA-2014-116904   9/23/2014    SC-20095  ...   944995        5  46304755

[9988 rows x 9 columns]

In [38]:
df.sort_values(['sales'], ascending=True)

order_id  order_date customer_id  ...     sales quantity   profit
8466  US-2017-132031   4/23/2017    PN-18775  ...         3        1      105
4660  CA-2016-111409   9/18/2016    PO-18850  ...         3        1      105
2259  CA-2015-117611   11/8/2015    MZ-17335  ...         5        1       24
1884  CA-2017-154718   1/19/2017    DL-12865  ...         6        2       21
9557  CA-2016-129280    5/3/2016    SM-20905  ...         6        2       21
...              ...         ...         ...  ...       ...      ...      ...
4263  US-2016-131611   11/5/2016    EP-13915  ...   9566648        7 -2250976
3     US-2015-108966  10/11/2015    SO-20335  ...   9575775        5  -383031
8775  CA-2015-133585    3/1/2015    CM-12715  ...  12279984        6  -361176
5052  CA-2015-141243    1/3/2015    AH-10465  ...  13523976        9 -4375404
399   CA-2016-108987    9/8/2016    AG-10675  ...  23962656        4 -3171528

[9988 rows x 9 columns]

In [ ]:
#Mengurutkan data frame dengan lebih dari 1 kolom
df.sort_values(['profit', 'quantity'], ascending=[True, False])

order_id  order_date customer_id  ...    sales quantity    profit
683   US-2017-168116   11/4/2017    GT-14635  ...   799998        4 -38399904
9768  CA-2014-169019   7/26/2014    LF-17185  ...  2177584        8 -37018928
4988  US-2017-122714   12/7/2017    HG-14965  ...   188999        5 -29294845
3149  CA-2015-147830  12/15/2015    NF-18385  ...  1799994        2 -26399912
9633  CA-2015-116638   1/28/2015    JH-15985  ...  4297644       13 -18623124
...              ...         ...         ...  ...      ...      ...       ...
4274  US-2016-107440   4/16/2016    BS-11365  ...   909993        7  23659818
1085  US-2016-143819    3/1/2016    KD-16270  ...   489993        7  24009657
6517  CA-2017-138289   1/16/2017    AR-10540  ...   544396        4  25042216
7662  US-2016-140158   10/4/2016    DR-12940  ...   539991        9  25919568
4096  CA-2014-116904   9/23/2014    SC-20095  ...   944995        5  46304755

[9988 rows x 9 columns]

In [42]:
df.sort_values(['sales', 'profit'], ascending=[False, True])

order_id  order_date customer_id  ...     sales quantity   profit
399   CA-2016-108987    9/8/2016    AG-10675  ...  23962656        4 -3171528
5052  CA-2015-141243    1/3/2015    AH-10465  ...  13523976        9 -4375404
8775  CA-2015-133585    3/1/2015    CM-12715  ...  12279984        6  -361176
3     US-2015-108966  10/11/2015    SO-20335  ...   9575775        5  -383031
2403  US-2016-110170   9/27/2016    HM-14860  ...   9566648        7 -2250976
...              ...         ...         ...  ...       ...      ...      ...
6105  CA-2016-150007   9/11/2016    AS-10090  ...         6        2       21
9557  CA-2016-129280    5/3/2016    SM-20905  ...         6        2       21
2259  CA-2015-117611   11/8/2015    MZ-17335  ...         5        1       24
4660  CA-2016-111409   9/18/2016    PO-18850  ...         3        1      105
8466  US-2017-132031   4/23/2017    PN-18775  ...         3        1      105

[9988 rows x 9 columns]

**Menghapus Missing Data(NA)**

In [ ]:
df.dropna()

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9988 rows x 9 columns]

In [43]:
df.dropna()

order_id  order_date customer_id  ...    sales quantity  profit
0     CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1     CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2     CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3     US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4     US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164
...              ...         ...         ...  ...      ...      ...     ...
9983  CA-2014-110422   1/21/2014    TB-21400  ...    25248        3   41028
9984  CA-2017-121258   2/26/2017    DB-13060  ...     9196        2  156332
9985  CA-2017-121258   2/26/2017    DB-13060  ...   258576        2  193932
9986  CA-2017-121258   2/26/2017    DB-13060  ...      296        4    1332
9987  CA-2017-119914    5/4/2017    CC-12220  ...    24316        2   72948

[9988 rows x 9 columns]

### Fungsi-fungsi pada Data Frame

**Count**

In [ ]:
#Menghitung jumlah baris data dalam suatu kolom
df.count()

order_id        9988
order_date      9988
customer_id     9988
segment         9988
category        9988
sub_category    9988
sales           9988
quantity        9988
profit          9988
dtype: int64

In [44]:
df.count()

order_id        9988
order_date      9988
customer_id     9988
segment         9988
category        9988
sub_category    9988
sales           9988
quantity        9988
profit          9988
dtype: int64

**Value Counts**

In [ ]:
#Menghitung jumlah data per kelas dalam data kategori
df.segment.value_counts()

Consumer       5188
Corporate      3019
Home Office    1781
Name: segment, dtype: int64

In [45]:
df.segment.value_counts()

Consumer       5188
Corporate      3019
Home Office    1781
Name: segment, dtype: int64

**Sum**

In [ ]:
#Menjumlahkan data dalam satu kolom numerik
df.profit.sum()

1659387772

In [46]:
df.sales.sum()

1109427931

**Max**

In [ ]:
#Nilai maximum data dalam satu kolom numerik
df.profit.max()

46304755

In [47]:
df.sales.max()

23962656

**Min**

In [ ]:
#Nilai minimum data dalam satu kolom numerik
df.profit.min()

-38399904

In [48]:
df.sales.min()

3

**Mean**

In [ ]:
#Nilai rata-rata data dalam satu kolom numerik
df.profit.mean()

166138.1429715659

In [49]:
df.sales.mean()

111076.08440128154

**Isnull()**

In [ ]:
#Menghitung jumlah data yang mengandung NA dalam sebuah dataset
df.isnull().sum()

order_id        0
order_date      0
customer_id     0
segment         0
category        0
sub_category    0
sales           0
quantity        0
profit          0
dtype: int64

In [50]:
df.isnull().sum()

order_id        0
order_date      0
customer_id     0
segment         0
category        0
sub_category    0
sales           0
quantity        0
profit          0
dtype: int64

## Eksploratory Data Anlysis[EDA] Dasar

Exploratory Data Analysis (EDA) adalah bagian dari proses data science. EDA menjadi sangat penting sebelum melakukan feature engineering dan modeling karena dalam tahap ini kita harus memahami datanya terlebih dahulu.

Exploratory Data Analysis mengacu pada proses kritis dalam melakukan investigasi awal pada data untuk menemukan pola, untuk menemukan anomali, untuk menguji hipotesis dan untuk memeriksa asumsi dengan bantuan statistik ringkasan dan representasi grafis. Dengan melakukan EDA, kita dapat lebih memahami kondisi dataset yang kita miliki. 

### Mengakses dataset

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/jokoeliyanto/Kelas-Dasar-Pejuang-Data-2.0/main/Super-Store-Dataset.csv")
df.head() 

order_id  order_date customer_id  ...     sales quantity    profit
0  CA-2016-152156   11/8/2016    CG-12520  ...  261.9600        2   41.9136
1  CA-2016-152156   11/8/2016    CG-12520  ...  731.9400        3  219.5820
2  CA-2016-138688   6/12/2016    DV-13045  ...   14.6200        2    6.8714
3  US-2015-108966  10/11/2015    SO-20335  ...  957.5775        5 -383.0310
4  US-2015-108966  10/11/2015    SO-20335  ...   22.3680        2    2.5164

[5 rows x 9 columns]

In [51]:
df = pd.read_csv("https://raw.githubusercontent.com/Alifhidayatunnisa/Basic-Python-Programming-for-Data-Science/main/Super%20Store%20Dataset.csv")
df.head()

order_id  order_date customer_id  ...    sales quantity  profit
0  CA-2016-152156   11/8/2016    CG-12520  ...    26196        2  419136
1  CA-2016-152156   11/8/2016    CG-12520  ...    73194        3  219582
2  CA-2016-138688   6/12/2016    DV-13045  ...     1462        2   68714
3  US-2015-108966  10/11/2015    SO-20335  ...  9575775        5 -383031
4  US-2015-108966  10/11/2015    SO-20335  ...    22368        2   25164

[5 rows x 9 columns]

### Memahami dataset

Variabel-variabel yang terdapat pada dataset ini adalah sebagai berikut:
1. **order_id** = Nomor ID pembelian(nota)
2. **order_data** = Tanggal pembelian
3. **customer_id** = Nomor ID pembeli
4. **segment** = Segmen produk 
5. **category** = Kategori produk
6. **sub_category** = Sub kategori produk
7. **sales** = Total harga pembelian
8. **quantity** = Jumlah produk dibeli
9. **profit** = Keuntungan


### Dimensi Data

In [ ]:
df.shape

(9988, 9)

In [52]:
df.shape

(9988, 9)

Artinya kita memiliki data dengan 9 kolom dan 9988 baris

### Variabel Pada Dataset

In [ ]:
df.columns.values 

array(['order_id', 'order_date', 'customer_id', 'segment', 'category',
       'sub_category', 'sales', 'quantity', 'profit'], dtype=object)

In [53]:
df.columns.values

array(['order_id', 'order_date', 'customer_id', 'segment', 'category',
       'sub_category', 'sales', 'quantity', 'profit'], dtype=object)

In [ ]:
df.dtypes

order_id         object
order_date       object
customer_id      object
segment          object
category         object
sub_category     object
sales           float64
quantity          int64
profit          float64
dtype: object

In [54]:
df.dtypes

order_id        object
order_date      object
customer_id     object
segment         object
category        object
sub_category    object
sales            int64
quantity         int64
profit           int64
dtype: object

In [ ]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9988 entries, 0 to 9987
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   order_id      9988 non-null   object 
 1   order_date    9988 non-null   object 
 2   customer_id   9988 non-null   object 
 3   segment       9988 non-null   object 
 4   category      9988 non-null   object 
 5   sub_category  9988 non-null   object 
 6   sales         9988 non-null   float64
 7   quantity      9988 non-null   int64  
 8   profit        9988 non-null   float64
dtypes: float64(2), int64(1), object(6)
memory usage: 702.4+ KB


In [55]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9988 entries, 0 to 9987
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   order_id      9988 non-null   object
 1   order_date    9988 non-null   object
 2   customer_id   9988 non-null   object
 3   segment       9988 non-null   object
 4   category      9988 non-null   object
 5   sub_category  9988 non-null   object
 6   sales         9988 non-null   int64 
 7   quantity      9988 non-null   int64 
 8   profit        9988 non-null   int64 
dtypes: int64(3), object(6)
memory usage: 702.4+ KB


Kita dapat mengetahui tipe-tipe data masing-masing variabel dan nama-nama variabel dalam dataset

### Mengecek Missing Data

In [ ]:
df.isnull().sum()

order_id        0
order_date      0
customer_id     0
segment         0
category        0
sub_category    0
sales           0
quantity        0
profit          0
dtype: int64

In [56]:
df.isnull().sum()

order_id        0
order_date      0
customer_id     0
segment         0
category        0
sub_category    0
sales           0
quantity        0
profit          0
dtype: int64

Kolom-kolom dengan data kosong adalah: Age, Cabin, Embarked

### Melihat Statistik Data

In [ ]:
df.describe()

sales     quantity       profit
count  9988.000000  9988.000000  9988.000000
mean    221.952598     3.789147    27.105341
std     514.818043     2.225586   186.115016
min       0.444000     1.000000 -3839.990400
25%      17.248000     2.000000     1.730250
50%      54.376000     3.000000     8.652800
75%     209.814000     5.000000    29.345400
max    9892.740000    14.000000  4946.370000

In [57]:
df.describe()

sales     quantity        profit
count  9.988000e+03  9988.000000  9.988000e+03
mean   1.110761e+05     3.789147  1.661381e+05
std    5.038903e+05     2.225586  1.438964e+06
min    3.000000e+00     1.000000 -3.839990e+07
25%    2.512000e+03     2.000000  2.245000e+03
50%    1.102150e+04     3.000000  3.629500e+04
75%    5.390700e+04     5.000000  1.509840e+05
max    2.396266e+07    14.000000  4.630476e+07

## Eksploratory Data Analysis[EDA] Lanjut

Setelah EDA dasar, eksplorasi selanjutnya adalah berupa pencarian insight sesuai kebutuhan kita. Misal kita ingin menjawab sebuah pertanyaan seperti berikut:

***Segment Produk Apakah Yang memiliki Keuntungan Tertinggi?***

In [ ]:
#Melihat Segment Apa Saja Yang Ada
df.segment.value_counts()

Consumer       5188
Corporate      3019
Home Office    1781
Name: segment, dtype: int64

In [58]:
df.category.value_counts()

Office Supplies    6026
Furniture          2121
Technology         1841
Name: category, dtype: int64

In [ ]:
#Memilih masing-masing segmen
df_Consumer=df[df['segment']=='Consumer']
df_Corporate=df[df['segment']=='Corporate']
df_Home_Office=df[df['segment']=='Home Office']

In [60]:
df_Office_Supplies=df[df['category']=='Office Supplies']
df_Furniture=df[df['category']=='Furniture']
df_Technology=df[df['category']=='Technology']

In [ ]:
df_Consumer.profit.sum()

128959.22080000001

In [ ]:
df_Corporate.profit.sum()

83579.158

In [ ]:
df_Home_Office.profit.sum()

58189.7681

In [63]:
df_Office_Supplies.profit.sum()

775994074

In [64]:
df_Furniture.profit.sum()

59222369

In [65]:
df_Technology.profit.sum()

824171329

Jadi, segmen dengan keuntungan total tertinggi adalah **Consumer**

Jadi, category dengan keuntungan total tertinggi adalah **Technology**